File: src/clewin/RayMarcher.java
package clewin;

/**
 * (c) Clewin Griffith 2013
 */
public class RayMarcher {

	private DistanceEstimator distanceEstimator;
	
	public RayMarcher(DistanceEstimator distanceEstimator) {
		this.distanceEstimator = distanceEstimator;
	}
	
	public double march(Ray ray) {
		
		double delt = 0.01;
		double eps = 0.001;
		double mint = 0.0;
		double maxt = 10.0;
		
		int maxSteps = 64;
		double totalDistance = 0.0;
		
		for(int step=0; step<maxSteps; step++)
		{
			Point p = Vector.pointOnLine(ray.start, ray.direction, totalDistance);
			double d = distanceEstimator.distanceFrom(p);
			if(d<eps)
			{
				//System.out.println("hit");
				//return 1.0;
				//double depth=Vector.length(ray.start, p);
				//System.out.println(depth);
				//return 0.5; //(10.0-totalDistance)/5;
				return 1.0-((double)step/maxSteps);
			}
			totalDistance+=d;
		}
		//System.out.println("miss");
		return 0.0;
	}
			
	
}

File: src/clewin/RayMarcher.scala
package clewin

/**
 * (c) Clewin Griffith 2013
 */
class RayMarcher(val distanceEstimator: DistanceEstimator) extends GeometryIntersector {

  def findIntersection(ray: Ray): RayIntersection = {
    val eps: Double = 0.001
    val maxSteps: Int = 128
    var totalDistance: Double = 0.0
    var mint = 0.01
    var maxt = 100.0
    var minDistanceFactor = 1.0
    var minDistance = 1000.0
      var t: Double = mint
      while (t < maxt) {

          val p: Point = Util.pointOnLine(ray.start, ray.direction, t)
          val d: Double = distanceEstimator.distanceFrom(p)
          minDistance = Math.min(minDistance, d)
          minDistanceFactor = Math.min(minDistanceFactor, d/t)
          if (d < eps) {
            return RayIntersection(p,d,ray, minDistanceFactor)// (step.asInstanceOf[Double] / maxSteps)
          }
         // totalDistance += d
          t += d
          //step += 1

      }

    return RayIntersection(null, minDistance,ray, minDistanceFactor)
  }

}